File: codeconvert_ai/input2.f90
subroutine fit_uspline(n,xs,p,q,j,en,FF)
use pietc, only: o3
use pmat2, only: ldltb, ltdlbv
integer,                intent(in ):: n
real(dp),dimension(  n),intent(in ):: xs,p
real(dp),dimension(  n),intent(out):: q,j
real(dp),               intent(out):: en
logical,                intent(out):: FF

integer                   :: i,ip
real(dp)                  :: x,x2,sa,sb,ccc,xcmsx2
real(dp),dimension(n-1)   :: difp,sumq,cpp,cqp
real(dp),dimension(n,-1:0):: qq 

FF=F
if(n<1)stop 'In fit_uspline; size of data array must be positive'
if(n==1)then; q=0; j=0; en=0; return; endif

do i=2,n
   if(xs(i-1)>=xs(i)) then
      FF=T
      write(41,*) 'In fit_uspline; xs data must increase strictly monotonically'
      return
   end if
enddo

qq=0 

do i=1,n-1
   ip=i+1
   difp(i)=p(ip)-p(i)
   x2=xs(ip)-xs(i); x=o2*x2
   xcmsx2=o3*x**3*2


   ccc=2/x
   cpp(i)=u1/xcmsx2 
   cqp(i)=-difp(i)*x/xcmsx2 
   qq(i,0)=qq(i,0)+ccc; qq(ip,-1)=qq(ip,-1)+1/x; qq(ip,0)=qq(ip,0)+ccc
enddo
q(1:n-1)=-cqp; q(n)=0
q(2:n)=q(2:n)-cqp


call ldltb(n,1,qq) 
call ltdlbv(n,1,qq,q)
sumq=q(1:n-1)+q(2:n) 

en=o2*(dot_product(difp**2,cpp)+dot_product(sumq,cqp))

sb=0
do i=1,n-1
   ip=i+1
   x=o2*(xs(ip)-xs(i))
   xcmsx2=o3*x**3*2
   sa=(x*sumq(i)-difp(i))/xcmsx2
   j(i)=sa-sb
   sb  =sa
enddo
j(n)=-sb 
end subroutine fit_uspline
  
subroutine int_uspline(n,xs,p,q, m)
integer,              intent(in ):: n
real(dp),dimension(n),intent(in ):: xs,p,q
real(dp),dimension(n),intent(out):: m

real(dp),parameter:: u3o2=3*o2
real(dp):: a,b,c,d,e,t2,t3,t4,x,pa,pd,qa,qd
integer :: i,ip
e=u0
do i=1,n-1
   ip=i+1
   x=(xs(ip)-xs(i))*o2 
   t2=x*x/2
   t3=t2*x/3
   t4=t3*x/4
   pa=(p(ip)+p(i))*o2
   pd=(p(ip)-p(i))*o2/x
   qa=(q(ip)+q(i))*o2
   qd=(q(ip)-q(i))*o2/x
   c=qd
   a=pa-c*t2
   d=(qa-pd)*u3o2/t2
   b=qa-d*t2
   m(i)=e+a*x-b*t2+c*t3-d*t4
   e=e+2*(a*x+c*t3)
enddo
m(n)=e
end subroutine int_uspline

subroutine eval_tspline(n,xs,p,q, x,y)
integer,              intent(in ):: n
real(dp),dimension(n),intent(in ):: xs,p,q
real(dp),             intent(in ):: x
real(dp),             intent(out):: y

integer :: ia,ib
real(dp):: xr,xh,pm,qm,qah,qbh,qxh,qdh,shh,chh,sh,ch,xcmsh,shm,chm,shhm,chhm

if(x<=xs(1))then; xr=x-xs(1); y=p(1)+q(1)*expm( xr); return; endif
if(x>=xs(n))then; xr=x-xs(n); y=p(n)-q(n)*expm(-xr); return; endif
do ib=2,n
   if(xs(ib)<=xs(ib-1))cycle
   if(xs(ib)>=x)exit 
enddo
ia=ib-1
xh=(xs(ib)-xs(ia))*o2 
xr=x-xs(ia)-xh 
pm=(p(ia)+p(ib))*o2 
qm=(p(ib)-p(ia))/(xh*2) 
qah=q(ia)*o2; qbh=q(ib)*o2
qxh=qah+qbh-qm 
qdh=qbh-qah   
shh=sinh(xh);   chh=cosh(xh)
sh =sinh(xr);    ch=cosh(xr)
shm=sinhm(xr);  chm=coshm(xr)
shhm=sinhm(xh); chhm=coshm(xh)
xcmsh=xcms(xh)
qdh=qdh/shh; qxh=qxh/xcmsh 
y=pm+xr*qm +qdh*(chm-chhm) +  qxh*(xh*shm-xr*shhm)
end subroutine eval_tspline
   

subroutine eval_tsplined(n,xs,p,q, x,y,dydx)



integer,              intent(in ):: n
real(dp),dimension(n),intent(in ):: xs,p,q
real(dp),             intent(in ):: x
real(dp),             intent(out):: y,dydx

integer :: ia,ib
real(dp):: xr,xh,pm,qm,qah,qbh,qxh,qdh,shh,chh,sh,ch,xcmsh,shm,chm,shhm,chhm,&
           qemxr

if(x<=xs(1))then
   xr=x-xs(1); qemxr=q(1)*expm( xr); y=p(1)+qemxr; dydx=qemxr+q(1); return
endif
if(x>=xs(n))then
   xr=x-xs(n); qemxr=q(n)*expm(-xr); y=p(n)-qemxr; dydx=qemxr+q(n); return
endif
do ib=2,n
   if(xs(ib)<=xs(ib-1))cycle 
   if(xs(ib)>=x)exit    
enddo
ia=ib-1
xh=(xs(ib)-xs(ia))*o2 
xr=x-xs(ia)-xh          
pm=(p(ia)+p(ib))*o2   
qm=(p(ib)-p(ia))/(xh*2) 
qah=q(ia)*o2;  qbh=q(ib)*o2
qxh=qah+qbh-qm 
qdh=qbh-qah    
shh=sinh(xh);   chh=cosh(xh)
sh =sinh(xr);   ch=cosh(xr)
shm=sinhm(xr);  chm=coshm(xr)
shhm=sinhm(xh); chhm=coshm(xh)
xcmsh=xcms(xh)
qdh=qdh/shh; qxh=qxh/xcmsh 
y=pm+xr*qm +qdh*(chm-chhm)  +  qxh*(xh*shm-xr*shhm)
dydx=qm+qdh*sh +qxh*(xh*chm-shhm)
end subroutine eval_tsplined


subroutine eval_tsplinedd(n,xs,p,q, x,y,dydx,ddydxx)



integer,              intent(in ):: n
real(dp),dimension(n),intent(in ):: xs,p,q
real(dp),             intent(in ):: x
real(dp),             intent(out):: y,dydx,ddydxx

integer :: ia,ib
real(dp):: xr,xh,pm,qm,qah,qbh,qxh,qdh,shh,chh,sh,ch,xcmsh,shm,chm,shhm,chhm,&
           qemxr

if(x<=xs(1))then
   xr=x-xs(1); qemxr=q(1)*expm( xr); y=p(1)+qemxr; dydx=qemxr+q(1)
   ddydxx=dydx; return
endif
if(x>=xs(n))then
   xr=x-xs(n); qemxr=q(n)*expm(-xr); y=p(n)-qemxr; dydx=qemxr+q(n)
   ddydxx=-dydx; return
endif
do ib=2,n
   if(xs(ib)<=xs(ib-1))cycle 
   if(xs(ib)>=x)exit    
enddo
ia=ib-1
xh=(xs(ib)-xs(ia))*o2 
xr=x-xs(ia)-xh          
pm=(p(ia)+p(ib))*o2   
qm=(p(ib)-p(ia))/(xh*2) 
qah=q(ia)*o2;  qbh=q(ib)*o2
qxh=qah+qbh-qm 
qdh=qbh-qah    
shh=sinh(xh);   chh=cosh(xh)
sh =sinh(xr);   ch=cosh(xr)
shm=sinhm(xr);  chm=coshm(xr)
shhm=sinhm(xh); chhm=coshm(xh)
xcmsh=xcms(xh)
qdh=qdh/shh; qxh=qxh/xcmsh 
y=pm+xr*qm +qdh*(chm-chhm)  +  qxh*(xh*shm-xr*shhm)
dydx=qm+qdh*sh +qxh*(xh*chm-shhm)
ddydxx=qdh*ch +qxh*xh*sh
end subroutine eval_tsplinedd


subroutine eval_tsplineddd(n,xs,p,q, x,y,dydx,ddydxx,dddydxxx)



integer,              intent(in ):: n
real(dp),dimension(n),intent(in ):: xs,p,q
real(dp),             intent(in ):: x
real(dp),             intent(out):: y,dydx,ddydxx,dddydxxx

integer :: ia,ib
real(dp):: xr,xh,pm,qm,qah,qbh,qxh,qdh,shh,chh,sh,ch,xcmsh,shm,chm,shhm,chhm,&
           qemxr

if(x<=xs(1))then
   xr=x-xs(1); qemxr=q(1)*expm( xr); y=p(1)+qemxr; dydx=qemxr+q(1)
   ddydxx=dydx; dddydxxx=dydx; return
endif
if(x>=xs(n))then
   xr=x-xs(n); qemxr=q(n)*expm(-xr); y=p(n)-qemxr; dydx=qemxr+q(n)
   ddydxx=-dydx; dddydxxx=dydx; return
endif
do ib=2,n
   if(xs(ib)<=xs(ib-1))cycle 
   if(xs(ib)>=x)exit    
enddo
ia=ib-1
xh=(xs(ib)-xs(ia))*o2 
xr=x-xs(ia)-xh          
pm=(p(ia)+p(ib))*o2   
qm=(p(ib)-p(ia))/(xh*2) 
qah=q(ia)*o2;  qbh=q(ib)*o2
qxh=qah+qbh-qm 
qdh=qbh-qah    
shh=sinh(xh);   chh=cosh(xh)
sh =sinh(xr);   ch=cosh(xr)
shm=sinhm(xr);  chm=coshm(xr)
shhm=sinhm(xh); chhm=coshm(xh)
xcmsh=xcms(xh)
qdh=qdh/shh; qxh=qxh/xcmsh 
y       =pm+xr*qm +qdh*(chm-chhm)  +  qxh*(xh*shm-xr*shhm)
dydx    =      qm +qdh*sh          +  qxh*(xh*chm-   shhm)
ddydxx  =          qdh*ch          +  qxh* xh*sh
dddydxxx=          qdh*sh          +  qxh* xh*ch
end subroutine eval_tsplineddd


subroutine eval_itspline(n,xs, p,q,m,  x,y)



integer,              intent(in ):: n
real(dp),dimension(n),intent(in ):: xs,p,q,m
real(dp),             intent(in ):: x
real(dp),             intent(out):: y

real(dp):: a,b,c,d,t2,xh,shx,chmx,shmx,chmmx,xcmsx,xr,pa,pd,qa,qd
integer :: ia,ib

if(x<=xs(1))then; xr=x-xs(1); y=     p(1)*xr+q(1)*expmm( xr); return; endif
if(x>=xs(n))then; xr=x-xs(n); y=m(n)+p(n)*xr+q(n)*expmm(-xr); return; endif
do ib=2,n
   if(xs(ib)<=xs(ib-1))cycle 
   if(xs(ib)>=x)exit 
enddo
ia=ib-1
xh=(xs(ib)-xs(ia))*o2   
shx =sinh (xh)
chmx=coshm(xh)
xcmsx=xcms(xh)
xr=x-xs(ia)-xh          
pa=(p(ib)+p(ia))*o2
pd=(p(ib)-p(ia))*o2/xh
qa=(q(ib)+q(ia))*o2
qd=(q(ib)-q(ia))*o2/shx

c=qd
a=pa-c*chmx
d=(qa-pd)*xh/xcmsx
b=qa-d*chmx

t2=xr**2/2
shmx =sinhm (xr)
chmmx=coshmm(xr)
y=m(ia)+a*xr+b*t2+c*shmx+d*chmmx
end subroutine eval_itspline


subroutine eval_uspline(n,xs,p,q, x,y)









integer,              intent(in ):: n
real(dp),dimension(n),intent(in ):: xs,p,q
real(dp),             intent(in ):: x
real(dp),             intent(out):: y

integer :: ia,ib
real(dp):: xr,xh,pm,qm,qah,qbh,qxh,qdh,shh,chh,sh,ch,xcmsh,shm,chm,shhm,chhm

if(x<=xs(1))then; xr=x-xs(1); y=p(1)+q(1)*xr; return; endif
if(x>=xs(n))then; xr=x-xs(n); y=p(n)+q(n)*xr; return; endif
do ib=2,n
   if(xs(ib)<=xs(ib-1))cycle 
   if(xs(ib)>=x)exit 
enddo
ia=ib-1
xh=(xs(ib)-xs(ia))*o2   
xr=x-xs(ia)-xh          
pm=(p(ia)+p(ib))*o2     
qm=(p(ib)-p(ia))/(xh*2) 
qah=q(ia)*o2;  qbh=q(ib)*o2
qxh=qah+qbh-qm 
qdh=qbh-qah    
shh=xh;        chh=u1
sh =xr;        ch =u1
shm =xr**3/6;  chm =xr**2*o2
shhm=xh**3/6;  chhm=xh**2*o2
xcmsh=xh**3/3
qdh=qdh/shh; qxh=qxh/xcmsh 
y=pm+xr*qm +qdh*(chm-chhm) +  qxh*(xh*shm-xr*shhm)

end subroutine eval_uspline
   

File: codeconvert_ai/input2.py
def fit_uspline(n, xs, p):
    import numpy as np
    q = np.zeros(n)
    j = np.zeros(n)
    en = 0
    FF = False
    if n < 1:
        raise ValueError('In fit_uspline; size of data array must be positive')
    if n == 1:
        q = np.zeros(n)
        j = np.zeros(n)
        en = 0
        return q, j, en, FF
    for i in range(1, n):
        if xs[i-1] >= xs[i]:
            FF = True
            print('In fit_uspline; xs data must increase strictly monotonically')
            return q, j, en, FF
    qq = np.zeros((n, 2))
    for i in range(n-1):
        ip = i + 1
        difp = p[ip] - p[i]
        x2 = xs[ip] - xs[i]
        x = x2 / 2
        xcmsx2 = (x ** 3) * 2 / 3
        ccc = 2 / x
        cpp = 1 / xcmsx2
        cqp = -difp * x / xcmsx2
        qq[i, 0] += ccc
        qq[ip, -1] += 1 / x
        qq[ip, 0] += ccc
        q[0:n-1] = -cqp
        q[n-1] = 0
        q[1:n] -= cqp
    ldltb(n, 1, qq)
    ltdlbv(n, 1, qq, q)
    sumq = q[0:n-1] + q[1:n]
    en = (np.dot(difp ** 2, cpp) + np.dot(sumq, cqp)) / 2
    sb = 0
    for i in range(n-1):
        ip = i + 1
        x = (xs[ip] - xs[i]) / 2
        xcmsx2 = (x ** 3) * 2 / 3
        sa = (x * sumq[i] - difp[i]) / xcmsx2
        j[i] = sa - sb
        sb = sa
    j[n-1] = -sb
    return q, j, en, FF

def int_uspline(n, xs, p, q):
    import numpy as np
    m = np.zeros(n)
    u3o2 = 3 / 2
    e = 0
    for i in range(n-1):
        ip = i + 1
        x = (xs[ip] - xs[i]) / 2
        t2 = x ** 2 / 2
        t3 = t2 * x / 3
        t4 = t3 * x / 4
        pa = (p[ip] + p[i]) / 2
        pd = (p[ip] - p[i]) / (2 * x)
        qa = (q[ip] + q[i]) / 2
        qd = (q[ip] - q[i]) / (2 * x)
        c = qd
        a = pa - c * t2
        d = (qa - pd) * u3o2 / t2
        b = qa - d * t2
        m[i] = e + a * x - b * t2 + c * t3 - d * t4
        e = e + 2 * (a * x + c * t3)
    m[n-1] = e
    return m

def eval_tspline(n, xs, p, q, x):
    import numpy as np
    ia = 0
    ib = 0
    xr = 0
    xh = 0
    pm = 0
    qm = 0
    qah = 0
    qbh = 0
    qxh = 0
    qdh = 0
    shh = 0
    chh = 0
    sh = 0
    ch = 0
    xcmsh = 0
    shm = 0
    chm = 0
    shhm = 0
    chhm = 0
    if x <= xs[0]:
        xr = x - xs[0]
        y = p[0] + q[0] * np.exp(xr)
        return y
    if x >= xs[n-1]:
        xr = x - xs[n-1]
        y = p[n-1] - q[n-1] * np.exp(-xr)
        return y
    for ib in range(1, n):
        if xs[ib] <= xs[ib-1]:
            continue
        if xs[ib] >= x:
            break
    ia = ib - 1
    xh = (xs[ib] - xs[ia]) / 2
    xr = x - xs[ia] - xh
    pm = (p[ib] + p[ia]) / 2
    qm = (p[ib] - p[ia]) / (2 * xh)
    qah = q[ia] / 2
    qbh = q[ib] / 2
    qxh = qah + qbh - qm
    qdh = qbh - qah
    shh = np.sinh(xh)
    chh = np.cosh(xh)
    sh = np.sinh(xr)
    ch = np.cosh(xr)
    shm = np.sinh(xr) / 6
    chm = np.cosh(xr) / 2
    shhm = np.sinh(xh) / 6
    chhm = np.cosh(xh) / 2
    xcmsh = (xh ** 3) / 3
    qdh = qdh / shh
    qxh = qxh / xcmsh
    y = pm + xr * qm + qdh * (chm - chhm) + qxh * (xh * shm - xr * shhm)
    return y

def eval_tsplined(n, xs, p, q, x):
    import numpy as np
    ia = 0
    ib = 0
    xr = 0
    xh = 0
    pm = 0
    qm = 0
    qah = 0
    qbh = 0
    qxh = 0
    qdh = 0
    shh = 0
    chh = 0
    sh = 0
    ch = 0
    xcmsh = 0
    shm = 0
    chm = 0
    shhm = 0
    chhm = 0
    if x <= xs[0]:
        xr = x - xs[0]
        qemxr = q[0] * np.exp(xr)
        y = p[0] + qemxr
        dydx = qemxr + q[0]
        return y, dydx
    if x >= xs[n-1]:
        xr = x - xs[n-1]
        qemxr = q[n-1] * np.exp(-xr)
        y = p[n-1] - qemxr
        dydx = qemxr + q[n-1]
        return y, dydx
    for ib in range(1, n):
        if xs[ib] <= xs[ib-1]:
            continue
        if xs[ib] >= x:
            break
    ia = ib - 1
    xh = (xs[ib] - xs[ia]) / 2
    xr = x - xs[ia] - xh
    pm = (p[ib] + p[ia]) / 2
    qm = (p[ib] - p[ia]) / (2 * xh)
    qah = q[ia] / 2
    qbh = q[ib] / 2
    qxh = qah + qbh - qm
    qdh = qbh - qah
    shh = np.sinh(xh)
    chh = np.cosh(xh)
    sh = np.sinh(xr)
    ch = np.cosh(xr)
    shm = np.sinh(xr) / 6
    chm = np.cosh(xr) / 2
    shhm = np.sinh(xh) / 6
    chhm = np.cosh(xh) / 2
    xcmsh = (xh ** 3) / 3
    qdh = qdh / shh
    qxh = qxh / xcmsh
    y = pm + xr * qm + qdh * (chm - chhm) + qxh * (xh * shm - xr * shhm)
    dydx = qm + qdh * sh + qxh * (xh * chm - shhm)
    return y, dydx

def eval_tsplinedd(n, xs, p, q, x):
    import numpy as np
    ia = 0
    ib = 0
    xr = 0
    xh = 0
    pm = 0
    qm = 0
    qah = 0
    qbh = 0
    qxh = 0
    qdh = 0
    shh = 0
    chh = 0
    sh = 0
    ch = 0
    xcmsh = 0
    shm = 0
    chm = 0
    shhm = 0
    chhm = 0
    qemxr = 0
    if x <= xs[0]:
        xr = x - xs[0]
        qemxr = q[0] * np.exp(xr)
        y = p[0] + qemxr
        dydx = qemxr + q[0]
        ddydxx = dydx
        return y, dydx, ddydxx
    if x >= xs[n-1]:
        xr = x - xs[n-1]
        qemxr = q[n-1] * np.exp(-xr)
        y = p[n-1] - qemxr
        dydx = qemxr + q[n-1]
        ddydxx = -dydx
        return y, dydx, ddydxx
    for ib in range(1, n):
        if xs[ib] <= xs[ib-1]:
            continue
        if xs[ib] >= x:
            break
    ia = ib - 1
    xh = (xs[ib] - xs[ia]) / 2
    xr = x - xs[ia] - xh
    pm = (p[ib] + p[ia]) / 2
    qm = (p[ib] - p[ia]) / (2 * xh)
    qah = q[ia] / 2
    qbh = q[ib] / 2
    qxh = qah + qbh - qm
    qdh = qbh - qah
    shh = np.sinh(xh)
    chh = np.cosh(xh)
    sh = np.sinh(xr)
    ch = np.cosh(xr)
    shm = np.sinh(xr) / 6
    chm = np.cosh(xr) / 2
    shhm = np.sinh(xh) / 6
    chhm = np.cosh(xh) / 2
    xcmsh = (xh ** 3) / 3
    qdh = qdh / shh
    qxh = qxh / xcmsh
    y = pm + xr * qm + qdh * (chm - chhm) + qxh * (xh * shm - xr * shhm)
    dydx = qm + qdh * sh + qxh * (xh * chm - shhm)
    ddydxx = qdh * ch + qxh * xh * sh
    return y, dydx, ddydxx

def eval_tsplineddd(n, xs, p, q, x):
    import numpy as np
    ia = 0
    ib = 0
    xr = 0
    xh = 0
    pm = 0
    qm = 0
    qah = 0
    qbh = 0
    qxh = 0
    qdh = 0
    shh = 0
    chh = 0
    sh = 0
    ch = 0
    xcmsh = 0
    shm = 0
    chm = 0
    shhm = 0
    chhm = 0
    qemxr = 0
    if x <= xs[0]:
        xr = x - xs[0]
        qemxr = q[0] * np.exp(xr)
        y = p[0] + qemxr
        dydx = qemxr + q[0]
        ddydxx = dydx
        dddydxxx = dydx
        return y, dydx, ddydxx, dddydxxx
    if x >= xs[n-1]:
        xr = x - xs[n-1]
        qemxr = q[n-1] * np.exp(-xr)
        y = p[n-1] - qemxr
        dydx = qemxr + q[n-1]
        ddydxx = -dydx
        dddydxxx = dydx
        return y, dydx, ddydxx, dddydxxx
    for ib in range(1, n):
        if xs[ib] <= xs[ib-1]:
            continue
        if xs[ib] >= x:
            break
    ia = ib - 1
    xh = (xs[ib] - xs[ia]) / 2
    xr = x - xs[ia] - xh
    pm = (p[ib] + p[ia]) / 2
    qm = (p[ib] - p[ia]) / (2 * xh)
    qah = q[ia] / 2
    qbh = q[ib] / 2
    qxh = qah + qbh - qm
    qdh = qbh - qah
    shh = np.sinh(xh)
    chh = np.cosh(xh)
    sh = np.sinh(xr)
    ch = np.cosh(xr)
    shm = np.sinh(xr) / 6
    chm = np.cosh(xr) / 2
    shhm = np.sinh(xh) / 6
    chhm = np.cosh(xh) / 2
    xcmsh = (xh ** 3) / 3
    qdh = qdh / shh
    qxh = qxh / xcmsh
    y = pm + xr * qm + qdh * (chm - chhm) + qxh * (xh * shm - xr * shhm)
    dydx = qm + qdh * sh + qxh * (xh * chm - shhm)
    ddydxx = qdh * ch + qxh * xh * sh
    dddydxxx = qdh * sh + qxh * xh * ch
    return y, dydx, ddydxx, dddydxxx

def eval_itspline(n, xs, p, q, m, x):
    import numpy as np
    ia = 0
    ib = 0
    xr = 0
    xh = 0
    pm = 0
    qm = 0
    qah = 0
    qbh = 0
    qxh = 0
    qdh = 0
    shh = 0
    chh = 0
    sh = 0
    ch = 0
    xcmsh = 0
    shm = 0
    chm = 0
    shhm = 0
    chhm = 0
    if x <= xs[0]:
        xr = x - xs[0]
        y = p[0] * xr + q[0] * np.expmm(xr)
        return y
    if x >= xs[n-1]:
        xr = x - xs[n-1]
        y = m[n-1] + p[n-1] * xr + q[n-1] * np.expmm(-xr)
        return y
    for ib in range(1, n):
        if xs[ib] <= xs[ib-1]:
            continue
        if xs[ib] >= x:
            break
    ia = ib - 1
    xh = (xs[ib] - xs[ia]) / 2
    xr = x - xs[ia] - xh
    pm = (p[ib] + p[ia]) / 2
    qm = (p[ib] - p[ia]) / (2 * xh)
    qah = q[ia] / 2
    qbh = q[ib] / 2
    qxh = qah + qbh - qm
    qdh = qbh - qah
    shh = xh
    chh = 1
    sh = xr
    ch = 1
    shm = xr ** 3 / 6
    chm = xr ** 2 / 2
    shhm = xh ** 3 / 6
    chhm = xh ** 2 / 2
    xcmsh = xh ** 3 / 3
    qdh = qdh / shh
    qxh = qxh / xcmsh
    y = m[ia] + pm * xr + qxh * (xh * shm - xr * shhm)
    return y

def eval_uspline(n, xs, p, q, x):
    import numpy as np
    ia = 0
    ib = 0
    xr = 0
    xh = 0
    pm = 0
    qm = 0
    qah = 0
    qbh = 0
    qxh = 0
    qdh = 0
    shh = 0
    chh = 0
    sh = 0
    ch = 0
    xcmsh = 0
    shm = 0
    chm = 0
    shhm = 0
    chhm = 0
    if x <= xs[0]:
        xr = x - xs[0]
        y = p[0] + q[0] * xr
        return y
    if x >= xs[n-1]:
        xr = x - xs[n-1]
        y = p[n-1] + q[n-1] * xr
        return y
    for ib in range(1, n):
        if xs[ib] <= xs[ib-1]:
            continue
        if xs[ib] >= x:
            break
    ia = ib - 1
    xh = (xs[ib] - xs[ia]) / 2
    xr = x - xs[ia] - xh
    pm = (p[ib] + p[ia]) / 2
    qm = (p[ib] - p[ia]) / (2 * xh)
    qah = q[ia] / 2
    qbh = q[ib] / 2
    qxh = qah + qbh - qm
    qdh = qbh - qah
    shh = xh
    chh = 1
    sh = xr
    ch = 1
    xcmsh = xh ** 3 / 3
    shm = xr ** 3 / 6
    chm = xr ** 2 / 2
    shhm = xh ** 3 / 6
    chhm = xh ** 2 / 2
    qdh = qdh / shh
    qxh = qxh / xcmsh
    y = pm + xr * qm + qdh * (chm - chhm) + qxh * (xh * shm - xr * shhm)
    return y



